File: algorithm/力扣热题100/20_有效的括号.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
    const st = []
    const map = {
        '(': ')',
        '[': ']',
        '{': '}',
    }
    for (let ch of s) {
        if (map[ch]) { // 左括号
            st.push(ch)
        } else {
            if (st.length === 0 || map[st.pop()] !== ch) return false
        }
    }
    return st.length === 0
};

File: algorithm/力扣热题100/20_有效的括号.go
package main

func isValid(s string) bool {
	m := map[byte]byte{
		'(': ')',
		'{': '}',
		'[': ']',
	}
	st := []byte{}
	for i := 0; i < len(s); i++ {
		if _, ok := m[s[i]]; ok {
			st = append(st, s[i])
		} else {
			if len(st) == 0 || m[st[len(st)-1]] != s[i] {
				return false
			}
			st = st[:len(st)-1]
		}
	}
	return len(st) == 0
}
